contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1024 @@
+    _removal = Removal(removalAddress);
+  }
+
+  /**
+   * Sets the duration in seconds that should be applied to schedules created on behalf of removals
+   * originating from the given methodology and methodology version.
+
+   * ##### Requirements:
+   *
+   * - Can only be used when the contract is not paused.
+   * - Can only be used when the caller has the `DEFAULT_ADMIN_ROLE`
+   */
+  function setRestrictionDurationForMethodologyAndVersion(
+    uint256 methodology,
+    uint256 methodologyVersion,
+    uint256 durationInSeconds
+  ) public whenNotPaused onlyRole(DEFAULT_ADMIN_ROLE) {
+    _methodologyAndVersionToScheduleDuration[methodology][
+      methodologyVersion
+    ] = RestrictionDuration(durationInSeconds, true);
+  }
+
+  /**
+   * Sets up a restriction schedule with parameters determined from the project id.
+   *
+   *
+   * ##### Requirements:
+   * - Can only be used when the contract is not paused.
+   * - Can only be used when the caller has the `SCHEDULE_CREATOR_ROLE` role
+   */
+  function createSchedule(uint256 projectId)
+    external
+    whenNotPaused
+    onlyRole(SCHEDULE_CREATOR_ROLE)
+  {
+    if (!_scheduleIdToScheduleStruct[projectId].exists) {
+      _createSchedule(projectId);
+    }
+  }
+
+  /**
+   * This function is triggered when BridgedPolygonNORI is sent to this contract
+   *
+   * @dev Sending BridgedPolygonNORI to this contract triggers the tokensReceived hook defined by the ERC-777 standard
+   * because this contract is a registered ERC777 tokens recipient.
+   *
+   * [See here for more](
+   * https://github.com/ethereum/EIPs/blob/master/EIPS/eip-777.md#erc777tokensrecipient-and-the-tokensreceived-hook)
+   */
+  function tokensReceived(
+    address,
+    address from,
+    address,
+    uint256 amount,
+    bytes calldata userData,
+    bytes calldata operatorData
+  ) external override {
+    if (!(_msgSender() == address(_bridgedPolygonNori))) {
+      revert TokenSenderNotBPNORI();
+    }
+    if (!(from == address(_market) || hasRole(DEFAULT_ADMIN_ROLE, from))) {
+      revert InvalidBpNoriSender({account: from});
+    }
+
+    uint256 removalId = abi.decode(userData, (uint256));
+    _depositFor(removalId, amount, userData, operatorData);
+  }
+
+  /**
+   * Claim released tokens and withdraw them to `recipient` address.
+   *
+   * @dev This function burns `amount` of `RestrictedNORI` for the given schedule id
+   * and transfers `amount` of `BridgedPolygonNORI` from the `RestrictedNORI` contract's
+   * balance to `recipient`'s balance.
+   *
+   * Enforcement of the availability of claimable tokens
+   * for the `_burn` call happens in `_beforeTokenTransfer`
+   *
+   * ##### Requirements:
+   *
+   * - Can only be used when the contract is not paused.
+   */
+  function withdrawFromSchedule(
+    address recipient,
+    uint256 scheduleId,
+    uint256 amount
+  ) external returns (bool) {
+    super._burn(_msgSender(), scheduleId, amount);
+    _bridgedPolygonNori.send(
```

Comment:
Just to be extra paranoid about the reentrancy risk here can this send call go at the bottom after the counters are updated?  I think having it after the burn is probably sufficient, but doesn't hurt to be extra paranoid.  If recipient is a contract it will get an `ERC777TokensReceived` callback from which it could maliciously call withdraw again.

Reply:
yeah definitely... I'm still planning to go through all the slither issues and rearrange the order of things to try to get rid of the reentrancy warnings. wanted to do that last when all of the other logic is settled with all this review churn